whispers/core/utils.py

```
@@ -0,0 +1,216 @@
+
+    if not file.is_file():
+        raise TypeError(f"{file.as_posix()} is not a file")
+
+
+def truncate_all_space(value: str) -> str:
+    """Replace multiple space characters by a single space character"""
+    if not value:
+        return ""
+
+    return re.sub(r"\s+", " ", value)
+
+
+def strip_string(value: str) -> str:
+    """Strips leading and trailing quotes and spaces"""
+    if not value:
+        return ""
+
+    return str(value).strip(" '\"\n\r\t")
+
+
+def simple_string(value: str) -> str:
+    """Returns a simplified value for loose comparison"""
+    if not value:
+        return ""
+
+    value = strip_string(value)  # Remove quotes
+    value = value.rstrip("\\")  # Remove trailing backslashes
+    value = value.lower()  # Lowercase
+    value = re.sub(r"[^a-z0-9]", "_", value.strip())  # Simplify
+
+    return value
+
+
+def similar_strings(a: str, b: str) -> float:
+    """Returns similarity coefficient between two strings"""
+    a = simple_string(a).replace("_", "")
+    b = simple_string(b).replace("_", "")
+
+    return ratio(a, b)
+
+
+def is_ascii(data: str) -> bool:
+    """Checks if given data is printable text"""
+    if isinstance(data, bytes):
+        try:
+            data = data.decode("utf-8")
+        except Exception:
+            return False
+
+    if not isinstance(data, (str, int)):
+        return False
+
+    for ch in str(data):
+        if ch not in string.printable:
+            return False
+
+    return True
+
+
+def is_base64(data: str) -> bool:
+    """Checks if given data is base64-decodable to text"""
+    if not isinstance(data, str):
+        return False
+
+    try:
+        b64decode(data).decode("utf-8")
+        return True
+
+    except Exception:
+        return False
+
+
+def is_base64_bytes(data: str) -> bool:
+    """Checks if given data is base64-decodable to bytes"""
+    try:
+        return b64decode(data) != b""
+
+    except Exception:
+        return False
```

Capturing bare exceptions might be dangerous, we might be hiding errors. Can we just capture the ones we expect on these cases ex` binascii.Error`?

---

There are too many exceptions that could happen here. It could be wrongly guessing something as base64 and trying to decode it, or decode something that is usually base64 but not that time. 

All errors are automatically logged into `whispers.log` for later analysis.